docker_cli/_utils.py

```diff
@@ -0,0 +1,256 @@
+    """
+    package_mgrs = image.check_command_availability(get_supported_package_managers())
+
+    if package_mgrs:
+        package_mgr = get_package_manager(package_mgrs[0])
+    else:
+        ValueError("No recognized package manager found on parent image.")
+
+    return package_mgr
+
+
+def _url_reader_check(
+    image: Image, package_mgr: PackageManager
+) -> Tuple[Type[URLReader], str]:
+    """
+    Return the URL reader on a given image, and a string to install one if there is
+    none.
+
+    Parameters
+    ----------
+    base : Image
+        The image.
+    package_mgr : PackageManager
+        The images package manager.
+
+    Returns
+    -------
+    Tuple[Type[URLReader], str]
+        The installed URL reader and a string to install it if necessary.
+    """
+    url_programs = image.check_command_availability(get_supported_url_readers())
+    init_lines = ""
+    if url_programs:
+        url_program: Type[URLReader] = get_url_reader(url_programs[0])
+    else:
+        init_lines += (
+            "RUN "
+            + str(
+                package_mgr.generate_install_command(targets=["wget"], stringify=True)
+            )
+            + "\n"
+        )
+        url_program = get_url_reader("wget")
+
+    return url_program, init_lines
+
+
+def _image_command_check(
+    image_name: str,
+    configure: bool = False,
+    stdout: Optional[io.TextIOBase] = None,
+    stderr: Optional[io.TextIOBase] = None,
+) -> Tuple[PackageManager, Type[URLReader], str]:
+    """
+    Determine what relevant commands are present on the image.
+
+    Determines what package manager and URL reader are on the image, and
+    returns a set of initial lines to install a URL reader if none is present.
+
+    Parameters
+    ----------
+    image_name : str
+        The tag or ID by which the image can be found
+    configure : bool
+        Add configuration commands to the returned string if True. Defaults to False.
+    stdout : io.TextIOBase, optional
+        A file-like object to redirect stdout to. If None, no redirection will
+        occur. By default None
+    stderr : io.TextIOBase, optional
+        A file-like object to redirect stderr to. If None, no redirection will
+        occur. By default None
+
+    Returns
+    -------
+    Tuple[PackageManager, URLReader, str]
+        The Package Manager object, URL Reader object, and any install and
+        configuration lines required by the dockerfile.
+
+    Raises
+    ------
+    ValueError
+        If the parent image does not have a recognized package manager.
+    """
+
+    tag = f"{universal_tag_prefix()}-temp-{UniqueGenerator.generate(k=10)}"
+    try:
+        base: Image = Image.build(  # type: ignore
+            tag=tag,
+            dockerfile_string=f"FROM {image_name}\nRUN mkdir {tag}",
+            stdout=stdout,
+            stderr=stderr,
+        )
+    except CalledProcessError:
+        raise ValueError(f"Image not found: {image_name}")
```

Comment:
As an example, if I was to try to replace that logic with `base = Image(image_name)`, this code would give a `CalledProcessError` any time I try to run pytests for `oraclelinux:8.4` since that image is not present currently on the machine I'm testing with, but the current way would work without error.

Reply:
Ah, I see. The issue is that the base image might not already be available locally. So you create a dummy image that doesn't do much more than just inherit from the base image, which has the side-effect of implicitly performing a [`docker pull`](https://docs.docker.com/engine/reference/commandline/pull/) if the base image name didn't correspond to any local image. Makes sense now.

Keep in mind that this kind of reasoning should generally be documented in a comment alongside the code.

A few notes:

* It seems to me that this issue (the base image not being available locally) could hypothetically affect any of our functions that take an `Image` object as opposed to an `image_name` to inherit from. It just happened to be the case that this `image_command_check()` function is typically the first thing that gets called when a user provides a base image name/tag.
  
  Just as an example, if a user used the `dropin` command to try to spin up a container from an image that didn't exist locally, in that case the command would fail rather than automatically pulling down the image from a remote server.

  So the problem is not necessarily specific to `image_command_check()` -- there might be other places in the codebase where we might want to automatically ensure that a remote image is pulled if the expected image wasn't already available locally.

* Therefore, I would say that `image_command_check()` is currently violating the [single-responsibility principle](https://en.wikipedia.org/wiki/Single-responsibility_principle) -- it's doing two things: (1) ensuring that the image is available locally, and (2) checking for a valid package manager / URL fetcher.

* Note that this also means that the interface of `image_command_check()` is slightly inconsistent with similar functions. Instead of taking an `Image` object like other functions do, it takes an `str` object representing the image tag or ID.

So I'd suggest breaking this functionality off into a separate function that takes an `image_name` and ensures that the image is available locally by creating a simple dummy image that inherits from it. We can call this function before calling `image_command_check()` and might also decide to reuse it elsewhere in the codebase.

Then, I'd suggest changing the interface of `image_command_check()` to take an `Image` object rather than an `image_name` string.

Reply:
Also, I suspect that the `\nRUN mkdir {tag}` part of the Dockerfile is unnecessary -- I think it should be possible to just create a one-liner Dockerfile that just has a `FROM` command in order to ensure that the base image is available locally.

Reply:
> Also, I suspect that the \nRUN mkdir {tag} part of the Dockerfile is unnecessary
For the purposes of checking if the image exists it's not necessary, but when testing it helps to use this, since if multiple docker images are build simultaneously from the same base with the same dockerfile, they will be considered the same image and if one is deleted, the others will become empty references which will create errors when doing parallelized tests. (Another detail which I will comment into the code)

This line addresses that by creating a directory in each image that is unique, meaning that the state of each image is different and they will be handled differently by Docker. If one is deleted, the others are not.

It's a bit of a workaround. It would also be possible just to attempt `docker pull` on an image if it isn't detected already on the system.

Reply:
>if multiple docker images are build simultaneously from the same base with the same dockerfile, they will be considered the same image and if one is deleted, the others will become empty references which will create errors when doing parallelized tests.

I don't think this is true, unless there's some detail that I'm missing.

It's true that if you run `docker build` multiple times on the same Dockerfile with the same build context, you are really just getting one image with multiple tags rather than multiple distinct images.

But running `docker rmi` to delete one of those tags shouldn't cause the remaining tags to become broken -- it just removes that tag rather than deleting the underlying image. The remaining image tags should be usable as before until all of the tags have been removed.

For example, if I create this Dockerfile in my current directory:

```Dockerfile
FROM ubuntu:22.04
```

And build several images from it with different tags:

```console
$ docker build . -t tag1
$ docker build . -t tag2
$ docker build . -t tag3
```

I can see using `docker inspect` that these are all just the same image with multiple tags, rather than distinct images:

```console
$ docker inspect --format='{{.RepoTags}}' tag1
[tag1:latest tag2:latest tag3:latest]
```

Using `docker rmi` to delete one of the tags doesn't affect the remaining tags:

```console
$ docker rmi tag3
Untagged: tag3:latest
$ docker inspect --format='{{.RepoTags}}' tag1
[tag1:latest tag2:latest]
```

This is true even if I force removal using `docker rmi -f`

```console
$ docker rmi -f tag2
Untagged: tag2:latest
$ docker inspect --format='{{.RepoTags}}' tag1
[tag1:latest]
```

The underlying image still exists. I can spin up a container from it using `docker run`

```console
$ docker run --rm -it tag1 echo "hello world"
hello world
```

Maybe there was some other confounding factor that was causing issues with your parallelized tests?

Reply:
This may be me trying to fix an issue I wasn't understanding - there is an error that I get sometimes when I run `docker rmi` or more generally `python -m docker_cli remove`; in some circumstances I will get the following:

`Error response from daemon: conflict: unable to delete {IMAGE_ID} (must be forced) - image is referenced in multiple repositories`

With image-id being some hexadecimal ID. I end up having to force the image removal. I was unsuccessful in figuring out what was causing this bug and this code might have been an attempt to prevent failed removals on these temporary images.

Reply:
Implemented a fix -
`temp_image()` is a new context manager that handles the creation and deletion of this temporary image.
These functions have been removed from the `image_command_check()` function.

Reply:
>in some circumstances I will get the following:
>
>`Error response from daemon: conflict: unable to delete {IMAGE_ID} (must be forced) - image is referenced in multiple repositories`

Ahh OK I think I see what's going on. The behavior of `docker rmi` must be different when you pass it an image \*ID\* vs an image \*tag\*.

If you have some image that has multiple tags, and you do `docker rmi <one of the tags>`, then it just removes that tag but doesn't clean up the underlying image.

```console
$ docker build . -t tag1
[...]
$ docker build . -t tag2
[...]
$ docker build . -t tag3
[...]
$ docker rmi tag3
Untagged: tag3:latest
```

But if you try to remove the same multiple-tagged image via its image *ID* instead using `docker rmi <image ID>`, you get the error message because there are still multiple tags referencing that image.

```console
$ docker build . -t tag1
[...]
$ docker build . -t tag2
[...]
$ docker build . -t tag3
[...]
$ docker rmi 69930c179dc9
Error response from daemon: conflict: unable to delete 69930c179dc9 (must be forced) - image is referenced in multiple repositories
```

When trying to remove an image with a single tag, though, the behavior seems to be the same between removing-by-tag vs removing-by-ID. The tag is removed and the underlying image is cleaned up. There's no error message in either case.

```console
$ docker build . -t tag1
[...]
$ docker rmi tag1
Untagged: tag1:latest
Deleted: sha256:69930c179dc9a59a13ec45c3197e93d7cfd807703cd241b01ce7ebfda58e13d3
```

```console
$ docker build . -t tag1
[...]
$ docker rmi 69930c179dc9
Untagged: tag1:latest
Deleted: sha256:69930c179dc9a59a13ec45c3197e93d7cfd807703cd241b01ce7ebfda58e13d3
```

So, in our case, the `docker_cli remove` command is searching for the ID of each image to be removed and then attempting to delete those images by their ID(s). So we're getting an error when those images have multiple tags pointing at them.

I guess I still think that the best thing to do here is to expose the error message to the user so that they know to use the `-f/--force` flag to force removal in this case. But I'm definitely open to smarter ideas about how to handle this.

Reply:
Ah, I think we've got our conversations mixed up here - to describe the present state of the code:
- I've exposed errors now in `docker_cli remove` which implements the behavior you've described above. I marked that conversation resolved since it appeared to me that we are in agreement about that being at least an acceptable way for the code to work, and since it would probably take a substantial rework of the command, which is otherwise functioning fine, to fix the confusing statements it prints out. My one remaining thought about this is that maybe it'd be better for `docker_cli remove` to have a `--verbose` argument that prints all of those things out instead of requiring the user to silence it, since the information it produces is a bit confusing and unnecessary for most users.
- On the topic of temporary images, I have removed the `mkdir` command and created a context manager for these temp images. The manager spins up a temp image, yields a generator that provides it, and then deletes that image by tag when it is closed.

Reply:
Gotcha, that all sounds great. Replacing the `quiet` option with a `verbose` option seems like a good idea.